### ISSUE TITLE
Freezes Firefox

### ISSUE THREAD

**Issue by IgnaciodeNuevo** [matatk/landmarks#154]
Hi Matthew:

I've tryied on MacOS High Sierra v10.13.3 and Firefox v590b14 (64-bit) and completely freezes the browser.

Great work, cheers

**Comment by matatk** [comment]
Does this happen when you visit a particular page? If so, you'll need to send me the URL so I can try to reproduce the problem here.

I have tried the extension on Firefox Developer Edition 59, on the same version of macOS, and had no problems.

Does the problem occur for you on Firefox 58?

**Comment by IgnaciodeNuevo** [comment]
It started happening since v57. No special website, random ones.

**Comment by matatk** [comment]
Could you describe exactly what happens when you encounter the freeze? Does it happen when you activate the toolbar button, or use a keyboard shortcut? What happens when it freezes? Does it un-freeze by itself?

There is an option to turn on logging of messages relating to the dynamic refresh feature, which might be of use.

I've not encountered any such problems, and we will need to find a way for me to reproduce the ones you are having, or I won't be able to fix it. If you can think of any relevant information, please let me know.

Are you sure it's relating to the Landmarks extension specifically? I guess you've tried disabling all other extensions (and Landmarks) and seeing if it still happens?

**Comment by IgnaciodeNuevo** [comment]
Using Twitter via web or YouTube. When scrolling while the extension is activated in the toolbar section.

No other extension is activated at the same time.

**Comment by matatk** [comment]
Does the problem persist when Landmarks is uninstalled?

**Comment by IgnaciodeNuevo** [comment]
Nope.

**Comment by matatk** [comment]
Which version of the extension are you using?

There were some [performance issues in the past, which were fixed](https://github.com/matatk/landmarks/issues/127). The problems you're describing sound similar, but I have just spent some time trying to recreate them by scrolling around and resizing the window whilst looking at YouTube videos and Twitter feeds, and I can't make Firefox or Chrome freeze up.

In Landmarks' preferences, there is a setting "Show debugging information in the console". I enabled this and checked the (Browser Console) output whilst scrolling/resizing, and it looks like it is working as it should in terms of avoiding overloading the browser. (When lots of changes are being made to the content of a page, it waits some time before checking for landmarks. The time it waits increases exponentially if the page content is still changing. When the page stops changing, the timer gradually decreases.)

There could still be a problem, but if I'm unable to reproduce it there's very little I can do to fix it. Perhaps you could upload a video (e.g. to YouTube) of the problems you are having as that might give me some more information? This would be particularly helpful if you enable the debugging information and have that visible on-screen (Press <kbd>⌘</kbd>+<kbd>J</kbd> to open the Browser Console window) whilst you are experiencing the freezing problems.

My Mac is about 5 years old, has a 2.4GHz i7 processor, 8G of RAM, a dedicated GPU and macOS 10.13.3. Is your system similar? (I very-much doubt the GPU makes a difference to this specifically; it would affect scrolling performance overall, though, I imagine.)

**Comment by matatk** [comment]
It occurred to me that if you ever installed a beta version of the extension (e.g. on Firefox) you may now be stuck on the beta version, and not getting updated to the latest stable release (in which I believe these issues were fixed). The current stable release is 2.2.0.

**Comment by matatk** [comment]
Closing this for now. Over the past couple of years I've added the means (via the DevTools panel) to investigate how the mutation observer rate-limiting is working. I have not found any performance problems and have been unable to reproduce this particular one.

I am mindful that there *could* be performance issues, and try to ensure a good balance between performance and coverage of landmark region changes. I think the performance of the extension has improved, but if any specific situations arise that need work, I'd be happy to look into them.
